File: 08 CP Priority Queue/Heap Node Relation/d62_q3a_heap_node_relation.java
import java.util.Scanner;

public class d62_q3a_heap_node_relation
{
    public static boolean checkDescendant(int a, int b)
    {
        if(a == 0) return true;
        while(b > a)
        {
            b--;
            b /= 2;
        }
        if(a == b) return true;
        return false;
    }
    public static void main(String[] args)
    {
        Scanner Scanner = new Scanner(System.in);
        int n = Scanner.nextInt();
        int m = Scanner.nextInt();
        for(int i = 0; i < m; i++)
        {
            int a = Scanner.nextInt();
            int b = Scanner.nextInt();
            if(a == b)
            {
                System.out.println("a and b are the same node");
            }
            else if(a < b && checkDescendant(a,b))
            {
                System.out.println("a is an ancestor of b");
            }
            else if(b < a && checkDescendant(b,a))
            {
                System.out.println("b is an ancestor of a");
            }
            else
            {
                System.out.println("a and b are not related");
            }
        }
        Scanner.close();
    }
}

File: 08 CP Priority Queue/Heap Node Relation/d62_q3a_heap_node_relation.py
import sys
def checkDescendant(a:int, b:int) -> bool:
    if(a == 0):
        return True
    while(b > a):
        b = (b-1)//2
    if(a == b):
        return True
    return False

n,m = map(int, sys.stdin.readline().split())
for _ in range(m):
    a,b = map(int, sys.stdin.readline().split())
    if(a == b):
        sys.stdout.write("a and b are the same node\n")
    elif(a < b and checkDescendant(a,b)):
        sys.stdout.write("a is an ancestor of b\n")
    elif(b < a and checkDescendant(b,a)):
        sys.stdout.write("b is an ancestor of a\n")
    else:
        sys.stdout.write("a and b are not related\n")